docs/gen-versions.py

```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env python3
+
+import argparse
+import json
+import os
+import subprocess
+import sys
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        prog="gen-versions.py",
+        description="Generate Sphinx version switcher JSON file from git tags.",
+    )
+
+    parser.add_argument(
+        "-o",
+        "--out-file",
+        help="path to output file, including filename, for generated versions JSON file",
+        metavar="path",
+        default=None,
+    )
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+
+    completed = subprocess.run(["./gen-versions.sh"], capture_output=True)
```

Comment:
just thinking out loud. wouldn't it make sense for gen-versions to be a python file that this module could import?

Reply:
It was easier/faster to perform the tag filtering with a `git` invocation directly (feat. `comm`) than to attempt the same behavior with something like GitPython, whose interface is not entirely the same as git itself. I will grant that it's not pretty, but libraries like that sometimes have to fork and exec git under the hood anyway.

Reply:
I did, at one point (not sure if it's reflected in the history or not), have an entirely different Python-only script that used `GitPython` for this, but that tag-finding behavior worked differently than `version.sh` did. Specifically, it was always returning the newest tag regardless of which branch you were on, whereas `version.sh` is sensitive to your position relative to other tags in the history.

Reply:
😬